### DISCUSSION TITLE
dedup items in Vec<T> field

### DISCUSSION THREAD

**Discussion by nklhtv** [jonasbb/serde_with#477] [Q&A] [answered]
Hi,

How to deduplicate vector entries of a field?
I've tried to use FromInto<HashSet<T>> but that did not work out.
I've also tried to `impl<'de, T, U> DeserializeAs<'de, Vec<T>> for Dedup<U>` where Dedup is `struct Dedup<T: Hash + Eq>(PhantomData<T>);` but could not complete that either.

```rust
#[derive(Hash, PartialEq, Eq)]
struct A {
    id: String,
    name: String,
}

#[derive(From, Into)]
struct B(A);

impl Hash for B {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.0.id.hash(state);
    }
}

impl PartialEq for B {
    fn eq(&self, other: &Self) -> bool {
        self.0.id == other.0.id
    }
}

impl Eq for B {}

struct MyStruct {
    #[serde_as(as = "Dedup<B>")]
    a: Vec<A>,
}
```

Any suggestions?
Eventually i will need to sort the vec, but i cannot use HashSet or BTreeSet

**Comment by jonasbb** [comment]
> I've also tried to `impl<'de, T, U> DeserializeAs<'de, Vec<T>> for Dedup<U>` where Dedup is `struct Dedup<T: Hash + Eq>(PhantomData<T>);` but could not complete that either.

Could you show what you tried so far? And please also include the serialized form of your data and how you expect the deserialized form to look like.

> I've tried to use `FromInto<HashSet>` but that did not work out.

That is because there is no `impl<T> From<HashSet<T>> for Vec<T>`. You could work with `IntoIterator` and `FromIterator`, but there is nothing premade.

**Comment by nklhtv** [comment, marked as answer]
Hi @jonasbb You can check [this commit](https://github.com/Stremio/stremio-core/commit/6727866174637597cc54473c96e6ca06e9cb496b) to see my current implementation.
i want to deduplicate manifest catalogs by id & type when deserializing and serializing, while preserving the original order.
example: 
```
{catalogs: [{"id": "top", type: "movie", ... }, {"id": "top", type: "series", ...}, {"id": "top", type: "movie", ...}] }
```
must be deserialized as 
```
vec![ManifestCatalog { id: "top", r#type: "movie", ... }, ManifestCatalog { id: "top", r#type: "series", ... }]
```

**Reply by jonasbb** [reply]
I don't know what is not working with your code, but here are some notes from looking at it.

> `impl SerializeAs<ManifestCatalog> for ManifestCatalogUnique`

This means in order to serialize a `ManifestCatalog` you want to run the serialization logic of `ManifestCatalogUnique`. The `SerializeAs` does *not* act like a wrapper type, it only works to change the serialization logic.

There are multiple things I can see here: https://github.com/Stremio/stremio-core/blob/6727866174637597cc54473c96e6ca06e9cb496b/src/types/serde_as_ext.rs#L13-L35

You specify that `Vec<T>` should be mapped to `U` here `DeserializeAs<'de, Vec<T>> for Unique<U>`, but your trait bounds specify that `T` maps to `U`, because you write: `U: DeserializeAs<'de, T>,`. In the helper you also assume `Vec<T>` maps to `Vec<U>` (`DeserializeAsWrap<Vec<T>, Vec<U>>`), similar to the where clause, not the `impl`.
You probably want `DeserializeAs<'de, Vec<T>> for Unique<Vec<U>>`.
The same "layer"-confusion also applies to `SerializeAs`.

Your `Helper` looks exactly like `DeserializeAsWrap`, except more complicated due to a useless lifetime. You even use `#[serde(transparent)]`

You implement `Visitor` for `Helper`, but are never using it.

---

Since you need a special `PartialEq` implementation, but only during deserialization, I don't know if `serde_as` really does what you want/need.

I would probably simplify it to a series of steps.
Deserialize a `Vec<ManifestCatalog>`. Turn it into a `Vec<ManifestCatalogUnique>`. Then call your `.unique()` on it. Unwrap the unique layer and turn it back into a `Vec<ManifestCatalog>`. Return that from your `deserialize` function.

**Reply by nklhtv** [reply]
The current impl was actually "working" but i do not like the code so i am trying to improve it.
Followed your advices and did some [refactoring](https://github.com/Stremio/stremio-core/compare/catalog-pagination)
think of Unique<U> as something like UniqueVec<U>, thats why i didnt put <Vec<U>> in there
Do i use this library in the correct way? My plan is to add another struct called SortedBy<...> and then stack them like
```#[serde_as(as = "SortedBy<UniqueBy<ManifestCatalog, (String, String), ManifestCatalogUniqueByAdapter>,....>")] ```

**Reply by jonasbb** [reply]
I'm sorry. My comment about the type bounds is indeed entirely wrong. I seemed to have confused myself quite badly. I'm glad to hear that everything is working for you. The refactoring is quite more readable. I'm wondering if something like this might be worthwhile.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6c401f4f984395d5b7085f831dde4435

**Reply by nklhtv** [reply]
@jonasbb Thank you for the proposal! It looks a lot better now, i just [applied](https://github.com/Stremio/stremio-core/commit/c9a722bbb4fb0fa9382d41fec94f3784b736071c) it.
I was also trying to remove the need of cloning (.to_owned()) in the `impl UniqueVecAdapter` blocks, but couldn't figure out the lifetimes.
My idea was to return (&String, &String) but that might be impossible :/
Any ideas?

**Reply by jonasbb** [reply]
@nklhtv Maybe something like this? I don't know what constraints the `unique_by` function requires.

```rust
pub trait UniqueByAdapter<'a> {
    type Original;
    type Comparable: Eq + Hash;

    fn call(t: &'a Self::Original) -> Self::Comparable;
}

impl<'a> UniqueByAdapter<'a> for ManifestCatalogUniqueByAdapter {
    type Original = ManifestCatalog;
    type Comparable = (&'a str, &'a str);

    fn call(value: &'a Self::Original) -> Self::Comparable {
        (&value.id, &value.r#type)
    }
}
```

**Reply by nklhtv** [reply]
@jonasbb Thats the signature of [unique_by](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.unique_by)
Thats my attempt to integrate your [last proposal](https://github.com/Stremio/stremio-core/commit/9927175516dd36604540cd76458499c601e2fc59), but did not work out.
One of the errors is very wierd. idk how to solve it: 
```
the type parameter `U` is not constrained by the impl trait, self type, or predicates
  --> src/types/serde_as_ext.rs:32:13
```
the other error, from rust-analyser, is in the DeserializeAs impl:
```
cannot infer an appropriate lifetime for lifetime parameter `'a` due to conflicting requirements

note: expected `<A as types::serde_as_ext::UniqueVecAdapter<'de>>::Input`
         found `<A as types::serde_as_ext::UniqueVecAdapter<'_>>::Input`
```
because `item` param in the unique_by callback has shorter lifetime than 'de :/

Maybe i should just keep using the owned values in the Adapter

**Reply by nklhtv** [reply]
I will keep using the owned values in the Adapter for now. Thank you for your help @jonasbb
